File: tools/UnityMonoBridge/src/win32/Memory.cpp
#include "win32/Memory.h"

#include <Windows.h>

#include <algorithm>
#include <array>
#include <sstream>
#include <stdexcept>

#include "win32/Process.h"

namespace bridge::win32 {

namespace {

std::string FormatWin32Message(const char* message, DWORD error_code, Address address = 0, std::size_t size = 0)
{
    std::ostringstream stream;
    stream << message << " (win32=" << error_code;
    if (address != 0) {
        stream << ", address=0x" << std::hex << address << std::dec;
    }
    if (size != 0) {
        stream << ", size=" << size;
    }
    stream << ")";
    return stream.str();
}

} // namespace

Memory::Memory(const Process& process)
    : process_(process)
{
}

Address Memory::Allocate(std::size_t size, unsigned long protection) const
{
    const auto address = reinterpret_cast<Address>(
        VirtualAllocEx(process_.handle(), nullptr, size, MEM_COMMIT | MEM_RESERVE, protection));
    if (address == 0) {
        throw std::runtime_error(FormatWin32Message("failed to allocate remote memory", GetLastError(), 0, size));
    }

    return address;
}

unsigned long Memory::Protect(Address address, std::size_t size, unsigned long protection) const
{
    DWORD previous_protection = 0;
    if (!VirtualProtectEx(process_.handle(), reinterpret_cast<void*>(address), size, protection, &previous_protection)) {
        throw std::runtime_error(FormatWin32Message("failed to change remote memory protection", GetLastError(), address, size));
    }

    return previous_protection;
}

void Memory::Free(Address address) const
{
    if (address != 0) {
        VirtualFreeEx(process_.handle(), reinterpret_cast<void*>(address), 0, MEM_RELEASE);
    }
}

void Memory::Write(Address address, const void* buffer, std::size_t size) const
{
    SIZE_T written = 0;
    if (!WriteProcessMemory(process_.handle(), reinterpret_cast<void*>(address), buffer, size, &written) || written != size) {
        throw std::runtime_error(FormatWin32Message("failed to write remote memory", GetLastError(), address, size));
    }
}

void Memory::Read(Address address, void* buffer, std::size_t size) const
{
    SIZE_T read = 0;
    if (!ReadProcessMemory(process_.handle(), reinterpret_cast<const void*>(address), buffer, size, &read) || read != size) {
        throw std::runtime_error(FormatWin32Message("failed to read remote memory", GetLastError(), address, size));
    }
}

std::string Memory::ReadUtf8(Address address, std::size_t max_length) const
{
    if (address == 0) {
        return {};
    }

    std::string result;
    result.reserve((std::min<std::size_t>)(max_length, 256));

    constexpr std::size_t chunk_size = 128;
    Address cursor = address;
    while (result.size() < max_length) {
        std::array<char, chunk_size> chunk{};
        const auto remaining = (std::min)(chunk.size(), max_length - result.size());
        Read(cursor, chunk.data(), remaining);

        const auto terminator = std::find(chunk.begin(), chunk.begin() + static_cast<std::ptrdiff_t>(remaining), '\0');
        result.append(chunk.begin(), terminator);
        if (terminator != chunk.begin() + static_cast<std::ptrdiff_t>(remaining)) {
            break;
        }

        cursor += remaining;
    }

    return result;
}

std::u16string Memory::ReadUtf16(Address address, std::size_t char_count) const
{
    if (address == 0 || char_count == 0) {
        return {};
    }

    std::u16string result(char_count, u'\0');
    Read(address, result.data(), char_count * sizeof(char16_t));
    return result;
}

void Memory::Execute(Address entry_point) const
{
    HANDLE thread = CreateRemoteThread(
        process_.handle(),
        nullptr,
        0,
        reinterpret_cast<LPTHREAD_START_ROUTINE>(entry_point),
        nullptr,
        0,
        nullptr);
    if (thread == nullptr) {
        throw std::runtime_error(FormatWin32Message("failed to create remote thread", GetLastError(), entry_point));
    }

    WaitForSingleObject(thread, INFINITE);
    CloseHandle(thread);
}

} // namespace bridge::win32

File: src-tauri/src/infrastructure/native/memory.rs
use crate::infrastructure::native::process::ProcessHandle;
use std::sync::Arc;
use windows::Win32::Foundation::CloseHandle;
use windows::Win32::System::Diagnostics::Debug::{ReadProcessMemory, WriteProcessMemory};
use windows::Win32::System::Memory::{
    VirtualAllocEx, VirtualFreeEx, VirtualProtectEx, MEM_COMMIT, MEM_RELEASE, MEM_RESERVE,
    PAGE_READWRITE,
};
use windows::Win32::System::Threading::{CreateRemoteThread, WaitForSingleObject, INFINITE};

#[derive(Debug, Clone)]
pub struct RemoteAllocation {
    memory: RemoteMemory,
    pub address: usize,
    pub size: usize,
}

impl Drop for RemoteAllocation {
    fn drop(&mut self) {
        let _ = self.memory.free_address(self.address);
    }
}

#[derive(Debug, Clone)]
pub struct RemoteUtf8String {
    allocation: RemoteAllocation,
}

impl RemoteUtf8String {
    pub fn new(memory: &RemoteMemory, value: &str) -> Result<Self, String> {
        let mut bytes = value.as_bytes().to_vec();
        bytes.push(0);
        let allocation = memory.allocate(bytes.len(), PAGE_READWRITE.0)?;
        memory.write_bytes(allocation.address, &bytes)?;
        Ok(Self { allocation })
    }

    pub fn address(&self) -> usize {
        self.allocation.address
    }
}

#[derive(Debug, Clone)]
pub struct RemoteMemory {
    process: Arc<ProcessHandle>,
}

impl RemoteMemory {
    pub fn open(pid: u32) -> Result<Self, String> {
        let process = Arc::new(crate::infrastructure::native::process::open_process(pid)?);
        Ok(Self { process })
    }

    pub fn process(&self) -> &ProcessHandle {
        self.process.as_ref()
    }

    pub fn allocate(&self, size: usize, protection: u32) -> Result<RemoteAllocation, String> {
        let address = unsafe {
            VirtualAllocEx(
                self.process.raw(),
                None,
                size,
                MEM_COMMIT | MEM_RESERVE,
                windows::Win32::System::Memory::PAGE_PROTECTION_FLAGS(protection),
            )
        } as usize;

        if address == 0 {
            return Err(format!(
                "Failed to allocate remote memory for process {} (size={})",
                self.process.pid(),
                size
            ));
        }

        Ok(RemoteAllocation {
            memory: self.clone(),
            address,
            size,
        })
    }

    pub fn protect(&self, address: usize, size: usize, protection: u32) -> Result<u32, String> {
        let mut previous = windows::Win32::System::Memory::PAGE_PROTECTION_FLAGS(0);
        unsafe {
            VirtualProtectEx(
                self.process.raw(),
                address as _,
                size,
                windows::Win32::System::Memory::PAGE_PROTECTION_FLAGS(protection),
                &mut previous,
            )
        }
        .map_err(|error| {
            format!(
                "Failed to change process memory protection at 0x{address:X}: {}",
                error.message()
            )
        })?;
        Ok(previous.0)
    }

    pub fn free_address(&self, address: usize) -> Result<(), String> {
        unsafe { VirtualFreeEx(self.process.raw(), address as _, 0, MEM_RELEASE) }
            .map_err(|error| format!("Failed to free remote allocation at 0x{address:X}: {}", error.message()))?;
        Ok(())
    }

    pub fn read_bytes(&self, address: usize, size: usize) -> Result<Vec<u8>, String> {
        let mut buffer = vec![0u8; size];
        let mut bytes_read = 0;
        unsafe {
            ReadProcessMemory(
                self.process.raw(),
                address as _,
                buffer.as_mut_ptr() as _,
                size,
                Some(&mut bytes_read),
            )
        }
        .map_err(|error| format!("Failed to read process memory at 0x{address:X}: {}", error.message()))?;
        buffer.truncate(bytes_read);
        Ok(buffer)
    }

    pub fn read_value<T: Copy>(&self, address: usize) -> Result<T, String> {
        let bytes = self.read_bytes(address, std::mem::size_of::<T>())?;
        if bytes.len() != std::mem::size_of::<T>() {
            return Err(format!(
                "Failed to read complete value at 0x{address:X}: {} of {} bytes",
                bytes.len(),
                std::mem::size_of::<T>()
            ));
        }

        let mut value = std::mem::MaybeUninit::<T>::uninit();
        unsafe {
            std::ptr::copy_nonoverlapping(bytes.as_ptr(), value.as_mut_ptr() as *mut u8, bytes.len());
            Ok(value.assume_init())
        }
    }

    pub fn write_bytes(&self, address: usize, data: &[u8]) -> Result<(), String> {
        let mut bytes_written = 0;
        unsafe {
            WriteProcessMemory(
                self.process.raw(),
                address as _,
                data.as_ptr() as _,
                data.len(),
                Some(&mut bytes_written),
            )
        }
        .map_err(|error| format!("Failed to write process memory at 0x{address:X}: {}", error.message()))?;

        if bytes_written != data.len() {
            return Err(format!(
                "Incomplete process memory write at 0x{address:X}: wrote {} of {} bytes",
                bytes_written,
                data.len()
            ));
        }

        Ok(())
    }

    pub fn write_value<T: Copy>(&self, address: usize, value: &T) -> Result<(), String> {
        let bytes = unsafe {
            std::slice::from_raw_parts(value as *const T as *const u8, std::mem::size_of::<T>())
        };
        self.write_bytes(address, bytes)
    }

    pub fn read_utf8(&self, address: usize, max_length: usize) -> Result<String, String> {
        if address == 0 {
            return Ok(String::new());
        }

        let mut result = String::new();
        for index in 0..max_length {
            let value = self.read_value::<u8>(address + index)?;
            if value == 0 {
                break;
            }
            result.push(value as char);
        }
        Ok(result)
    }

    pub fn read_utf16(&self, address: usize, char_count: usize) -> Result<Vec<u16>, String> {
        if address == 0 || char_count == 0 {
            return Ok(Vec::new());
        }

        let bytes = self.read_bytes(address, char_count * std::mem::size_of::<u16>())?;
        if bytes.len() % 2 != 0 {
            return Err(format!("Invalid UTF-16 byte length {} at 0x{address:X}", bytes.len()));
        }

        Ok(bytes
            .chunks_exact(2)
            .map(|chunk| u16::from_le_bytes([chunk[0], chunk[1]]))
            .collect())
    }

    pub fn execute(&self, entry_point: usize) -> Result<(), String> {
        let thread = unsafe {
            CreateRemoteThread(
                self.process.raw(),
                None,
                0,
                Some(std::mem::transmute(entry_point)),
                None,
                0,
                None,
            )
        }
        .map_err(|error| format!("Failed to create remote thread at 0x{entry_point:X}: {}", error.message()))?;

        unsafe {
            WaitForSingleObject(thread, INFINITE);
            let _ = CloseHandle(thread);
        }

        Ok(())
    }
}